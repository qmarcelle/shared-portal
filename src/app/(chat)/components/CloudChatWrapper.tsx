'use client';

// CloudChatWrapper injects the Genesys Cloud Messenger SDK and manages Messenger-specific events.
// All script loading, Messenger events, and errors are logged for traceability and debugging.

import { useEffect } from 'react';
import { useChatStore } from '../stores/chatStore';

export default function CloudChatWrapper() {
  const { userData } = useChatStore();

  useEffect(() => {
    // Build chatSettings from env and userData
    window.chatSettings = {
      bootstrapUrl: process.env.NEXT_PUBLIC_GENESYS_BOOTSTRAP_URL || '',
      widgetUrl: process.env.NEXT_PUBLIC_GENESYS_WIDGET_URL || '',
      clickToChatJs: process.env.NEXT_PUBLIC_GENESYS_CLICK_TO_CHAT_JS || '',
      clickToChatEndpoint: process.env.NEXT_PUBLIC_CLICK_TO_CHAT_ENDPOINT || '',
      chatTokenEndpoint: process.env.NEXT_PUBLIC_CHAT_TOKEN_ENDPOINT || '',
      coBrowseEndpoint: process.env.NEXT_PUBLIC_COBROWSE_LICENSE_ENDPOINT || '',
      opsPhone: process.env.NEXT_PUBLIC_OPS_PHONE || '',
      opsPhoneHours: process.env.NEXT_PUBLIC_OPS_HOURS || '',
      ...userData,
    };
    console.debug('[CloudChatWrapper] chatSettings:', window.chatSettings);
  }, [userData]);

  useEffect(() => {
    // Load Genesys Cloud bootstrap and widget scripts
    const bootstrapUrl = window.chatSettings?.bootstrapUrl || '';
    const widgetUrl = window.chatSettings?.widgetUrl || '';
    if (!bootstrapUrl) return;
    const bootstrapScript = document.createElement('script');
    bootstrapScript.src = bootstrapUrl;
    bootstrapScript.async = true;
    bootstrapScript.onload = () => {
      console.debug('[CloudChatWrapper] Genesys bootstrap loaded');
      if (!widgetUrl) return;
      const widgetScript = document.createElement('script');
      widgetScript.src = widgetUrl;
      widgetScript.async = true;
      widgetScript.onload = () => {
        console.debug('[CloudChatWrapper] Genesys widgets loaded');
      };
      widgetScript.onerror = (e) => {
        console.error('[CloudChatWrapper] Failed to load widgets script', e);
      };
      document.body.appendChild(widgetScript);
    };
    bootstrapScript.onerror = (e) => {
      console.error('[CloudChatWrapper] Failed to load bootstrap script', e);
    };
    document.body.appendChild(bootstrapScript);

    // Provide openGenesysChat logic
    window.openGenesysChat = () => {
      if (window.Genesys) {
        try {
          window.Genesys('command', 'Messenger.open');
        } catch (e) {
          console.error(
            '[CloudChatWrapper] Error opening Genesys Messenger',
            e,
          );
        }
      }
    };

    return () => {
      // Cleanup scripts and chatSettings
      document.body.removeChild(bootstrapScript);
      if (widgetUrl) {
        const widgetScript = document.querySelector(
          'script[src="' + widgetUrl + '"]',
        );
        if (widgetScript) document.body.removeChild(widgetScript);
      }
      if (window.chatSettings) delete window.chatSettings;
      if (window.openGenesysChat) delete window.openGenesysChat;
      console.debug('[CloudChatWrapper] Cleanup complete');
    };
  }, [userData]);

  return null;
}
