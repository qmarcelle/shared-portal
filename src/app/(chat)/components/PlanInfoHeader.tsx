'use client';
import { useChatStore } from '@/app/(chat)/stores/chatStore';
import { Title } from '@/components/foundation/Title';

/**
 * Show current plan when multiple plans exist
 */
export default function PlanInfoHeader() {
  const { formInputs } = useChatStore();
  if (formInputs.length <= 1) return null;
  const plan = formInputs.find((f) => f.id === 'PLAN_ID');

  return (
    <Title
      className="plan-info-header"
      text={`You are chatting about plan ${plan?.value || ''}`}
    />
  );
}
