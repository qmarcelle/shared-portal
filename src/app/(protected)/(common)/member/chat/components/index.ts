/**
 * Component Exports
 * Centralizes all component exports following the new consolidated structure
 */

// Main Components
export { ChatTrigger } from './ChatTrigger';
export { ChatWidget } from './ChatWidget';
export { PlanInfoHeader } from './PlanInfoHeader';

// Error Handling
export { ChatErrorBoundary } from './shared/ChatErrorBoundary';

// Integration Components (to be consolidated)
export { GenesysScripts } from './GenesysScripts';
