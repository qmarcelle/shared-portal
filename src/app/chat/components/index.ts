// Core Components
export * from './core/ChatButton';
export * from './core/ChatErrorBoundary';
export * from './core/ChatHeader';
export * from './core/ChatWidget';

// Business Hours Components
export * from './business-hours/BusinessHoursNotification';

// Eligibility Components
export * from './eligibility/EligibilityCheck';

// Plan Switcher Components will be added here when implemented
// export * from './plan-switcher/PlanSwitcher';
