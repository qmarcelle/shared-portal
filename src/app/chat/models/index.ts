// Export all chat-related types
export type {
  CobrowseConfig,
  CobrowseInitResponse,
  CobrowseSession,
  CobrowseSessionResponse,
} from './cobrowse';

// Make sure the ChatMessage type is consistent by defining it inline here
export interface ChatMessage {
  id: string;
  text: string;
  sender: 'user' | 'agent' | 'bot';
  timestamp: Date;
}

export * from './plans';
export type {
  ChatEligibility,
  ChatPayload,
  ChatSession,
  ChatSessionJWT,
  ChatState,
  PlanSwitcherState,
} from './session';
export type {
  BusinessDay,
  BusinessHours,
  ChatConfig,
  ChatOption,
  CobrowseState,
  UserEligibility,
} from './types';
