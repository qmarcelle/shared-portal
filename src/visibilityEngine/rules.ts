export const MEMBER_ATTRIBUTES_LIST = [
  'active',
  'termed',
  'futureEffective',
  'selfFunded',
  'fullyInsured',
  'levelFunded',
  'delinquent',
  'subscriber',
  'commercial',
  'individual',
  'medical',
  'dental',
  'vision',
  'wellness',
  'wellnessOnly',
  'spendingAccounts',
  'externalSpendingAcct',
  'teladoc',
  'pharmacy',
  'amplifyHealth',
  'premiumHealth',
  'benefitBooklet',
  'employerProvidedBenefits',
  'fsaOnly',
  'terminated',
  'katieBeckNoBenefitsElig',
  'blueCare',
  'myPCPElig',
  'chipRewardsEligible',
  'blueHealthRewardsEligible',
  'bluePerksElig',
  'amplifyMember',
  'primary360Eligible',
  'mentalHealthSupport',
  'commercialFull',
  'individualSBCEligible',
  'medicareAdvantageGroupIndicator',
  'isCondensedExperience',
  'enRollEligible',
  'enableBenefitChange',
  'medicare',
  'enableCostTools',
  'dentalCostsEligible',
  'payMyPremiumElig',
  'mentalHealthSupport',
  'myStrengthCompleteEligible',
  'primary360Eligible',
  'hingeHealthEligible',
  'groupRenewalDateBeforeTodaysDate',
  'healthCoachElig',
  'indivEHBUser',
  'groupEHBUser',
  'isAmplifyMem',
  'cmEnable',
] as const;

export type VisibilityRule = (typeof MEMBER_ATTRIBUTES_LIST)[number];
export type VisibilityRules = {
  [key in VisibilityRule]?: boolean;
};
