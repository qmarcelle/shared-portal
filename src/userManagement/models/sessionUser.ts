import { VisibilityRules } from '@/visibilityEngine/rules';

export enum UserRole {
  MEMBER = 'MEM',
  PERSONAL_REP = 'PR',
  AUTHORIZED_USER = 'AU',
  NON_MEM = 'NM',
}

export type SessionUser = {
  id: string;
  currUsr: {
    umpi: string;
    fhirId: string;
    role: UserRole;
    plan:
      | {
          memCk: string;
          sbsbCk: string;
          grgrCk: string;
          grpId: string;
          subId: string;
          fhirId: string;
          ntwkId: string;
        }
      | undefined;
  };
  rules?: string;
  vRules?: VisibilityRules;
  subscriberId?: string;
};
