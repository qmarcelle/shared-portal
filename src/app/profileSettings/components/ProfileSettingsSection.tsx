import { InfoCard } from '@/components/composite/InfoCard';
import { Column } from '@/components/foundation/Column';
import {
  communicationSettingsIcon,
  securityIcon,
  sharePermissionsIcon,
} from '@/components/foundation/Icons';
import { UserRole } from '@/userManagement/models/sessionUser';
import { checkPersonalRepAccess } from '@/utils/getRole';
import { isCommunicationSettingsEligible } from '@/visibilityEngine/computeVisibilityRules';
import { VisibilityRules } from '@/visibilityEngine/rules';

export type ProfileSettingsSectionProps = {
  visibilityRules?: VisibilityRules;
  userRole: UserRole | undefined;
};

export const ProfileSettingsSection = ({
  visibilityRules,
  userRole,
}: ProfileSettingsSectionProps) => {
  let profileSettingDetails;
  if (isCommunicationSettingsEligible(visibilityRules)) {
    profileSettingDetails = [
      {
        label: 'Communication Settings',
        description: 'Update your alert preferences.',
        iconName: communicationSettingsIcon,
        link: '/member/profile/communication',
      },
      {
        label: 'Security Settings',
        description: 'Change your password and edit your account security.',
        iconName: securityIcon,
        link: '/member/profile/security',
      },

      {
        label: 'Sharing & Permissions',
        description: 'View or edit access to plan information.',
        iconName: sharePermissionsIcon,
        link: '/member/profile/accountsharing',
      },
    ];
  } else {
    profileSettingDetails = [
      {
        label: 'Sharing & Permissions',
        description: 'View or edit access to plan information.',
        iconName: sharePermissionsIcon,
        link: '/member/profile/accountsharing',
      },
    ];
  }

  // Filter out settings based on user role
  if (userRole && !checkPersonalRepAccess(userRole)) {
    profileSettingDetails = profileSettingDetails.filter(
      (profileSetting) =>
        profileSetting.label !== 'Communication Settings' &&
        profileSetting.label !== 'Security Settings',
    );
  }

  return (
    <Column>
      {profileSettingDetails.map((item) => {
        return (
          <InfoCard
            key={item.label}
            label={item.label}
            icon={item.iconName}
            body={item.description}
            link={item.link}
          />
        );
      })}
    </Column>
  );
};
