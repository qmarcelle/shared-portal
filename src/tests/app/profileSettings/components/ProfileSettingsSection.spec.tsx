import { ProfileSettingsSection } from '@/app/profileSettings/components/ProfileSettingsSection';
import { UserRole } from '@/userManagement/models/sessionUser';
import { VisibilityRules } from '@/visibilityEngine/rules';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';

let vRules: VisibilityRules = {};
const renderUI = (vRules: VisibilityRules) => {
  return render(
    <ProfileSettingsSection visibilityRules={vRules} userRole={undefined} />,
  );
};
jest.mock('next/navigation', () => ({
  redirect: jest.fn(),
}));
jest.mock('src/auth', () => ({
  auth: jest.fn(() =>
    Promise.resolve({
      user: {
        currUsr: {
          role: UserRole.PERSONAL_REP,
        },
      },
    }),
  ),
}));
describe('Profile Settings Card Component', () => {
  beforeEach(() => {
    vRules = {};
  });

  it('should render UI correctly for Active Members', () => {
    vRules.terminated = false;
    vRules.futureEffective = false;
    const component = renderUI(vRules);
    screen.getByText('Communication Settings');
    screen.getByText('Update your alert preferences.');
    screen.getByText('Security Settings');
    screen.getByText('Change your password and edit your account security.');
    screen.getByText('Sharing & Permissions');
    screen.getByText('View or edit access to plan information.');
    expect(component.baseElement).toMatchSnapshot();
  });

  it('should render UI correctly for Terminated and Future Effective members', () => {
    vRules.terminated = true;
    vRules.futureEffective = false;
    const component = renderUI(vRules);
    screen.getByText('Sharing & Permissions');
    screen.getByText('View or edit access to plan information.');
    expect(component.baseElement).toMatchSnapshot();
  });

  it('should hide the if user role is PERSONAL_REP', async () => {
    render(<ProfileSettingsSection userRole={UserRole.PERSONAL_REP} />);
    expect(
      screen.queryByText('Communication Settings'),
    ).not.toBeInTheDocument();
    expect(screen.queryByText('Security Settings')).not.toBeInTheDocument();
  });
});
