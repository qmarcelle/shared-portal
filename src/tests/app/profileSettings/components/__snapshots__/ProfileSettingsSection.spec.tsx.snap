// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Settings Card Component should render UI correctly for Active Members 1`] = `
<body>
  <div>
    <div
      class="flex flex-col"
    >
      <a
        href="/communicationSettings"
      >
        <div
          class="card-elevated small-section"
        >
          <div
            class="flex flex-row"
          >
            <img
              alt="link"
              class="size-10"
              data-nimg="1"
              decoding="async"
              height="40"
              loading="lazy"
              src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
              style="color: transparent;"
              width="40"
            />
            <div
              class="flex flex-col ml-4"
            >
              <div
                class="flex flex-row link-row-head"
              >
                Communication Settings
                <span
                  style="display: block; width: 8px; min-width: 8px; height: 1px; min-height: 1px;"
                />
              </div>
              <p
                class="body-1 body-1"
              >
                Update your alert preferences.
              </p>
            </div>
          </div>
        </div>
      </a>
      <a
        href="/security"
      >
        <div
          class="card-elevated small-section"
        >
          <div
            class="flex flex-row"
          >
            <img
              alt="link"
              class="size-10"
              data-nimg="1"
              decoding="async"
              height="40"
              loading="lazy"
              src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
              style="color: transparent;"
              width="40"
            />
            <div
              class="flex flex-col ml-4"
            >
              <div
                class="flex flex-row link-row-head"
              >
                Security Settings
                <span
                  style="display: block; width: 8px; min-width: 8px; height: 1px; min-height: 1px;"
                />
              </div>
              <p
                class="body-1 body-1"
              >
                Change your password and edit your account security.
              </p>
            </div>
          </div>
        </div>
      </a>
      <a
        href="/sharingPermissions"
      >
        <div
          class="card-elevated small-section"
        >
          <div
            class="flex flex-row"
          >
            <img
              alt="link"
              class="size-10"
              data-nimg="1"
              decoding="async"
              height="40"
              loading="lazy"
              src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
              style="color: transparent;"
              width="40"
            />
            <div
              class="flex flex-col ml-4"
            >
              <div
                class="flex flex-row link-row-head"
              >
                Sharing & Permissions
                <span
                  style="display: block; width: 8px; min-width: 8px; height: 1px; min-height: 1px;"
                />
              </div>
              <p
                class="body-1 body-1"
              >
                View or edit access to plan information.
              </p>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</body>
`;

exports[`Profile Settings Card Component should render UI correctly for Terminated and Future Effective members 1`] = `
<body>
  <div>
    <div
      class="flex flex-col"
    >
      <a
        href="/sharingPermissions"
      >
        <div
          class="card-elevated small-section"
        >
          <div
            class="flex flex-row"
          >
            <img
              alt="link"
              class="size-10"
              data-nimg="1"
              decoding="async"
              height="40"
              loading="lazy"
              src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
              style="color: transparent;"
              width="40"
            />
            <div
              class="flex flex-col ml-4"
            >
              <div
                class="flex flex-row link-row-head"
              >
                Sharing & Permissions
                <span
                  style="display: block; width: 8px; min-width: 8px; height: 1px; min-height: 1px;"
                />
              </div>
              <p
                class="body-1 body-1"
              >
                View or edit access to plan information.
              </p>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</body>
`;
